provider/lxd/provider.go

```diff
@@ -166,25 +169,53 @@ func (p *environProvider) getLocalHostAddress(ctx environs.FinalizeCloudContext)
 
 	// LXD itself reports the host:ports that it listens on.
 	// Cross-check the address we have with the values reported by LXD.
-	if err := lxdclient.EnableHTTPSListener(raw); err != nil {
+	if err := raw.EnableHTTPSListener(); err != nil {
 		return "", errors.Annotate(err, "enabling HTTPS listener")
 	}
-	cInfo, err := raw.GetConnectionInfo()
-	if err != nil {
-		return "", errors.Trace(err)
-	}
-	var found bool
-	for _, addr := range cInfo.Addresses {
-		if strings.HasPrefix(addr, hostAddress+":") {
-			hostAddress = addr
-			found = true
-			break
-		}
+
+	// The following retry mechanism is required for bionic, where the new
+	// lxd client doesn't propagate the EnableHTTPSListener quick enough to
+	// get the addresses or on the same existing local provider.
+
+	// connInfoAddresses is really useful for debugging, so let's keep that
+	// information around for the debugging errors.
+	var connInfoAddresses []string
+	errNotExists := errors.New("not-exists")
+	retryArgs := retry.CallArgs{
+		Clock: p.clock(),
+		IsFatalError: func(err error) bool {
+			return errors.Cause(err) != errNotExists
+		},
+		Func: func() error {
+			// Requesting a newLocalRawProvider forces a new connection, so
+			// that when we GetConnectionInfo it gets the required addresses.
+			raw, err := p.newLocalRawProvider()
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			cInfo, err := raw.GetConnectionInfo()
```

Comment:
Did you try ```GetServer``` here? 

It has ```Environment.Addresses``` on it. If the addresses turned up there, we wouldn't have to reconnect each time.